indexing-service/src/test/java/org/apache/druid/indexing/common/task/CompactionTaskTest.java

```diff
@@ -1447,6 +1452,7 @@ private void assertIngestionSchema(
             null,
             null,
             null,
+            null,
             null
```

Comment:
Do we need a `CompactionTaskTest` where `awaitSegmentAvailabilityTimeoutMillis` is non null

Reply:
I doubt it. I don't think I was writing this with the intention of having compaction tasks support this. It really wouldn't add value since compaction isn't materially changing the underlying data. But I guess since compaction is spinning up a parallel indexing task, we kinda need to consider it so I will have to check on this.

Reply:
Interesting point. I think there are some things we should think about first.

- It's true that currently compaction doesn't change the underlying data much, but it can make some changes such as filtering out some unnecessary dimensions or adding new metrics. You can also change the query granularity now. In the future, I can imagine that you can even transform your data using compaction with a new support for transformSpec.
- The compaction task is a bit special and different from other batch tasks in how it publishes segments. All other batch tasks can push segments in the middle of indexing, but should publish all those segments at the end of indexing. However, the compaction task can process each time chunk at a time when there is no change in segment granularity. In this case, it can publish segments whenever it finishes processing individual time chunk. It can also go through all time chunks even when there are some time chunks that it fails to compact. The final task status will be `FAILED` when it succeeds to compact only some time chunks but fails for others.
- Compacting datasources is usually not the single-shot type job. Rather, you would run multiple small compaction tasks over time as in auto compaction. In that case, you would want to know what time chunks are compacted and what are not, so that you can determine what result you can get when you query certain time chunks. For the compaction that is manually set up outside druid, tracking of individual compaction tasks could be useful for this purpose. However, for auto compaction, it won't provide much value since compaction tasks are submitted by the coordinator not users. So, we need another way such as adding a new coordinator API that returns such compaction status.

From these, we would probably want something similar but different for compaction from the one proposed here. I would suggest to do it in a different PR.

Reply:
@jihoonson when it comes to judging segment handoff, isn't a compaction task a parallel indexing task under the hood? I just tested a compaction task submit with tuning config and the new config for handoff set to 5 minutes. The task ran and logs indicate it confirmed handoff of newly created compacted segment. 

Should not allow the handoff to occur when it is a compact type task? I'm not sure how we would do that, since the TuningConfig is not modifiable as written and if the task(s) ran by the compact task are just parallel indexing tasks, I'm not sure how we would tell them not to honor the handoff wait config

Reply:
> @jihoonson when it comes to judging segment handoff, isn't a compaction task a parallel indexing task under the hood? I just tested a compaction task submit with tuning config and the new config for handoff set to 5 minutes. The task ran and logs indicate it confirmed handoff of newly created compacted segment.

It is the parallel index task under the hood, but [one compaction task can run multiple parallel indexing tasks](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java#L420-L447). With the changes in this PR, each of those parallel indexing tasks will wait for segment handoff independently. The handoff wait timeout will also apply separately for each parallel task. I think this is OK as long as it's documented, but what I meant is, maybe we could do this differently, such as waiting for segments to be handed off all together after running all parallel tasks. 

> Should not allow the handoff to occur when it is a compact type task? I'm not sure how we would do that, since the TuningConfig is not modifiable as written and if the task(s) ran by the compact task are just parallel indexing tasks, I'm not sure how we would tell them not to honor the handoff wait config

It seems reasonable to me to not support until we figure out what we want to do. An easy way to do so is making a copy of tuningConfig in compaction task but with `awaitSegmentAvailabilityTimeoutMillis` always being set to 0. We should document that compaction tasks don't wait for handoff if you want to do this.

Reply:
Given this discussion, can we add a guardrail that says if someone tries to issue a compaction task when `awaitSegmentAvailabilityTimeoutMillis` is non null and != 0 that this is not supported. I think that's the easiest way forward to get the PR merged and it's a nice UX because it tells users that what they are doing is untested. Similar comment for auto-compaction (and other ingestion tasks that I may have missed)

Reply:
> Given this discussion, can we add a guardrail that says if someone tries to issue a compaction task when `awaitSegmentAvailabilityTimeoutMillis` is non null and != 0 that this is not supported. I think that's the easiest way forward to get the PR merged and it's a nice UX because it tells users that what they are doing is untested. Similar comment for auto-compaction (and other ingestion tasks that I may have missed)

Are you expecting a job submit failure then?

Reply:
Yes, that's what I was imagining. That being said, I'm open to other options if you had something in mind.

I'm basing this on the principle that if I (the user) do something that Druid doesn't expect, Druid tells me I'm doing something unexpected, and failing fast is better than failing in some obscure way later on in the process.

A clear error message on the failed job like "`awaitSegmentAvailabilityTimeoutMillis ` is not supported for compaction. Please remove this from the spec and re-submit the compaction job." would tell users exactly how to fix the failed job - hence, the nice UX.

A future PR for this support with auto-compaction / compaction would be super cool though.

Reply:
I like this idea of it failing. I will work on changes as soon as I can, hopefully early this week.